File: src/main/java/com/hoi4utils/clausewitz/data/country/CountryTag.java
package com.hoi4utils.clausewitz.data.country;

import com.hoi4utils.clausewitz.script.AbstractPDX;
import com.hoi4utils.clausewitz.script.PDXScript;
import org.jetbrains.annotations.NotNull;

public final class CountryTag extends AbstractPDX<String> implements Comparable<CountryTag> {
    public static final CountryTag NULL_TAG = new CountryTag("###");
    public static final int COUNTRY_TAG_LENGTH = 3;         // standard country tag length (for a normal country tag)private final String tag;

    public CountryTag() {
        super("tag");
    }

    public CountryTag(String tag) {
        super("tag");
        set(tag);
    }

//    public CountryTag(String tag) {
//        if (tag == null || tag.isEmpty()) {
//            tag = NULL_TAG.tag;
//        }
//        this.tag = tag;
//    }
//
//
//    @Override
//    public String toString() {
//        return tag;
//    }
//
//    // todo make this adhere to equals implementation requirements
//    @Override
//    public boolean equals(Object other) {
//        if (other.getClass() == this.getClass()) {
//            return this.tag.equals(((CountryTag) other).tag);
//        } else if (other.getClass() == String.class) {
//            return this.tag.equals(other);
//        }
//
//        return false;
//    }
//
    @Override
    public int compareTo(@NotNull CountryTag o) {
        return this.get().compareTo(o.get());
    }

    @Override
    public boolean objEquals(PDXScript<?> other) {
        if (other instanceof CountryTag pdx) {
            return this.nodeEquals(pdx);
        }
        return false;
    }

//
//    public String tag() {
//        return tag;
//    }
//
//    @Override
//    public int hashCode() {
//        return Objects.hash(tag);
//    }

}

File: src/main/scala/com/hoi4utils/clausewitz/data/country/CountryTag.scala
package com.hoi4utils.clausewitz.data.country

import scala.collection.mutable.ListBuffer

object CountryTag extends Iterable[CountryTag] {
  val NULL_TAG = new CountryTag("###")
  val COUNTRY_TAG_LENGTH = 3 // standard country tag length (for a normal country tag)private final String tag;
  // scala... (this is null (????????????) if you dont use 'lazy')
  private lazy val _tagList: ListBuffer[CountryTag] = {
    //    println("Initializing _tagList")
    ListBuffer[CountryTag]()
  }

  def get(tag: String): CountryTag = {
    for (countryTag <- _tagList) {
      if (countryTag.get.equals(tag)) return countryTag
    }
    NULL_TAG
  }

  override def iterator: Iterator[CountryTag] = {
    _tagList.iterator
  }

  //  def tagList(): List[CountryTag] = _tagList
  def addTag(tag: CountryTag): Unit = {
    _tagList.addOne(tag)
  }
}

class CountryTag(val tag: String) extends Comparable[CountryTag] {
  CountryTag.addTag(this)
  
  def get: String = tag
  
  override def toString: String = {
    tag
  }
  
  override def equals(obj: Any): Boolean = {
    obj match {
      case other: CountryTag =>
        tag.equals(other.tag)
      case _ => false
    }
  }

  override def compareTo(o: CountryTag): Int = {
    tag.compareTo(o.tag)
  }
}
